added withdrawWithSignature function

francisldn opened the pull request:
Draft PR for the issue - Create a new pool function to accept individually signed claims #736

francisldn commented:
@dmosites following the discussion yesterday, I thought that instead of recovering the signer, we can just compare the signed hash vs the hash provided by Activate. If the hashes match, then the token amount will be transferred.

AboldUSER commented:
Awesome and a couple comments:

- The modifier MultiAdmin can be used instead of the require function, to align with similar use of the admins mapping in the contract

- I see a second mapping is created, hashClaimed. If we have two different claim mappings, is there a risk that a user can receive double claim - once from the merkle tree route and second from an admin calling this function? Or if the signed message route is for claims not on the merkle tree then never mind

-  This piece of code (hashClaimed[ethSignedMessageHash] = true;) can go before the safeTransfer call. I think it is just always good practice to avoid putting state changing pieces after calling an external address

- In the current function code, I don't think it is set up to take a signature, instead it is taking the hashed data portion that would normally then be signed. So in this case Activate or whoever the admin is wont be sending a signed messaged, instead just sending token, recipient and amount arguments and a hash of those arguments. If Activate is indeed providing a signed message, then we'd have to use the ECDSA.recover function on the getEthSignedMessageHash output. In the case that they are just sending hashed data, I don't see the point of sending the arguments and then the hash of those arguments, and then verify the arguments and hash in the contract as it is just redundant. 

- The withdrawWithSignature can only ever be called on a recipient, amount, and token once. Since we are mapping the hash of those arguments to a bool, we would never be able to claim that same token & amount for a recipient in the future. Not sure if this is expected, but otherwise then the signed message would need to also contain at least salt/nonce.

- I don't think true needs to be returned, as there is also an event being emitted. Maybe can return an amount value similar to the withdrawProtected function

francisldn commented:
Thank you @AboldUSER. All very valid points. I have made some changes to the function based on your feedback:
- used the same "claimed" mapping for withdrawWithSignature 
- added nonce to the hash - assuming nonce is recorded off-chain
- added signature as an argument
- used multiAdmin modifier
- state change before external call
- return amount (uint) instead of boolean

francisldn commented:
2 further changes:
- updated the IPool interface - adding withdrawWithSignature function
- added 2 additional tests for the withdrawWithSignature function in the integration.js file

dmosites commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,34 @@ contract Pool is IPool, Ownable {
+    address token,
+    uint256 amount,
+    address recipient
+  ) external override multiAdmin returns (uint256) {
```
Should be callable by anyone. Admins sign and distribute claims to recipients who can in turn call this function to withdraw.

dmosites commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,34 @@ contract Pool is IPool, Ownable {
+    // verify signed hash has not been claimed
+    require(!claimed[messageHash][recipient], "CLAIM_ALREADY_MADE");
+    // to verify hash is signed by an admin, ECDSA.recover will throw error if signer is not recoverable
+    bytes32 ethSignedMessageHash = ECDSA.toEthSignedMessageHash(messageHash);
```
Hash should be calculated inline from the token, amount, recipient, and a nonce. Otherwise, anyone with a valid hash and signature can transfer any amount of any token.

francisldn replied:
Hi @dmosites thanks for the comments, sorry I misunderstood the requirements. I have a further question, in order to recover the signer, we need a signed hash and a signature of one of the admins, should "signed hash" be one of the arguments of the function? Does activate produce a signed hash?

dmosites replied:
An admin produces an off-chain claim that includes claim parameters and a signature. The `ecrecover` solidity function takes a hash and signature parameters as arguments. Should be able to compute a hash from claim parameters at runtime similar to what's happening in Swap: https://github.com/airswap/airswap-protocols/blob/main/source/swap/contracts/Swap.sol#L208

francisldn replied:
I have modified the function so that it can now recover the signer from the hash and signature using ``ECDSA.recover()`` function. I didn't use ``ecrecover`` function because it allows for malleable (non-unique) signatures according to Openzeppelin.

dmosites commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,34 @@ contract Pool is IPool, Ownable {
+    // to verify hash is signed by an admin, ECDSA.recover will throw error if signer is not recoverable
+    bytes32 ethSignedMessageHash = ECDSA.toEthSignedMessageHash(messageHash);
+    address signer = ECDSA.recover(ethSignedMessageHash, signature);
+    require(signer == msg.sender, "NOT_VERIFIED");
```
This should require that signer is among the admins i.e. `admins[signer] == true`

francisldn replied:
ok this is modified accordingly.

dmosites commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,34 @@ contract Pool is IPool, Ownable {
+    address signer = ECDSA.recover(ethSignedMessageHash, signature);
+    require(signer == msg.sender, "NOT_VERIFIED");
+    // mark the hash of (signature, hash and nonce) and recipient as true
+    claimed[messageHash][recipient] = true;
```
If hash is calculated inline (including recipient, token, amount, nonce) as suggested above, this just needs to be `claimed[hash] = true`

francisldn replied:
ok, this is modified accordingly.

dmosites commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,34 @@ contract Pool is IPool, Ownable {
+  */
+  function withdrawWithSignature(
+    bytes memory signature,
+    bytes32 messageHash,
```
Should take a `nonce` argument and calculate the hash in the function body (see below).

francisldn replied:
``nonce`` added to the argument.

dmosites replied:
Do we need the `messageHash` argument if we're generating it in the function body?

francisldn replied:
We would need it for verification purpose? If a user enters the wrong amount or token address (intentionally or unintentionally) then the generated hash will not match with the ``messageHash``.

AboldUSER replied:
Generally, signed message implementations take the individual parameters and the signature as arguments. The signature contains the message hash anyways. 
If a user is inputting incorrect arguments in a function, then you would want the transaction to fail.

francisldn replied:
Ok, so I will remove the ``messageHash`` and ``require(messageHash == calculatedHash, "HASH_NOT_VERIFIED")`` as any wrong input will result in the ``ECDSA.recover`` transaction to fail, so there is no need to compare the hash. Am I understanding this correctly?

AboldUSER replied:
Yup, that is correct.

AboldUSER commented on source/pool/contracts/Pool.sol:
```diff
@@ -363,4 +368,41 @@ contract Pool is IPool, Ownable {
+    address token,
+    uint256 amount,
+    uint256 nonce
+  ) external override returns (uint256) {
```
Since this is an external, state-changing function, it should go above internal and/or pure/view functions as a styling best practice. I'd suggest moving it to just above the withdrawProtected function.

AboldUSER commented on source/pool/test/integration.js:
```diff
@@ -293,8 +294,58 @@ describe('Pool Integration Tests', () => {
     })
-  })
 
+    it('withdrawWithSignature does not revert with signer who is admin', async () => {
```
I think these should actually go in the unit test file, as these functions are not interacting with any other contract.
